### PR TITLE
turned on matrix-multiplication => matrix-vector multiplication always on if reduction-dim is contiguous

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120954
* #128938

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by shunting314 — APPROVED** [review]
Is a perf test necessary since this cause some mv decomposition happening out of the box?

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -229,11 +229,22 @@ def mm(self, input2):
...
+        return (self.unsqueeze(2) * input2.unsqueeze(0)).sum(dim=1)
+
+    def coordesc_or(cond):
+        return config.coordinate_descent_tuning or cond
+
+    if config.decompose_mm_to_mv:
+        gso = guard_size_oblivious
+        # Corresponds to BS=1 cases morally. If it's contiguous we can decompose it
+        # and generate efficiently kernels without autotuning. Otherwise we need coordinate descent tuning
+        if gso(self.shape[0] == 1) and coordesc_or(gso(input2.stride(0) == 1)):
```
Comment (eellison) [inline torch/_inductor/decomposition.py:242]: Feel like if it's an unbacked symint here we probably dont want to decompose to mv, since that could be any shape and potentially large regression.  I think we'd want definitely_true

**Review by eellison — COMMENTED** [review]
Have any benchmarks you can share ?

**Comment by Chillee** [comment]
This is the benchmark I ran
```
import torch._dynamo.config
torch._dynamo.config.cache_size_limit = int(1e9)
for transpose_field in [(False, False), (False, True), (True, False), (True, True)]:
    results = []
    random.seed(0)
    for _ in range(20):
        N = random.randint(2048//16, 16384//16) * 16
        M = random.randint(2048//16, 16384//16) * 16
        dtype=torch.bfloat16

        def bench(f, name=None):
            import time
            from triton.testing import do_bench

            us_per_iter = do_bench(lambda: f())*1000
            bw_gb = (1e6/us_per_iter) * N * M / 1e9
            # print(f"{name}: {bw_gb} GB/s")

            return bw_gb

        def f(x, W):
            return F.linear(x, W.to(x.dtype))

        cf = torch.compile(f, dynamic=False)

        W = torch.randn(N, M).to(dtype=dtype)
        x = torch.randn(1, M, dtype=torch.bfloat16)
        if transpose_field[0]:
            x = x.t().contiguous().t()
        if transpose_field[1]:
            W = W.t().contiguous().t()

        eager = bench(lambda: f(x, W))
        compiled = bench(lambda: cf(x, W))
        results.append(compiled/eager)

    results = sorted(results)
    print(results[0], results[len(results)//2], results[-1])
```

Results in (this is relative speedup of compiled over eager)
```
min median max
1.0231771168834158 1.1344278165713064 1.444073714086211
0.9978222386237598 1.000235305717646 1.0061505333908283
1.0249540539509723 1.1382784788329456 1.4267091339420284
0.9844501516653604 0.9995172693841768 1.0070561531493678
```

**Comment by Chillee** [comment]
@shunting314 tbh I think it's unlikely to affect our perf on our dashboard - not many matrix vector multiplications going on.

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -229,11 +229,28 @@ def mm(self, input2):
...
+        gso = guard_size_oblivious
+        # Corresponds to BS=1 cases morally. If it's contiguous we can decompose it
+        # and generate efficient kernels without autotuning. Otherwise we need coordinate descent tuning
+        if definitely_true(self.shape[0] == 1) and coordesc_or(
+            definitely_true(input2.stride(0) == 1)
+        ):
+            return mul_sum_decomp(self, input2)
+
+        if definitely_true(input2.shape[1] == 1) and coordesc_or(
+            definitely_true(self.stride(1) == 1)
```
Comment (eellison) [inline torch/_inductor/decomposition.py:250]: The stride observed during decomposition is not necessarily the stride during lowering. We might want to add a force contiguous node or something on the meta

Comment (Chillee) [reply torch/_inductor/decomposition.py:250]: imo, it's fine to leave that up to Inductor. Like, the biggest reason we want to force strides for extern kernels is that Inductor has no visibility into what the extern kernel is going to do, and so if we generate the wrong strides then from Inductor's perspective the kernels may be efficient, but we're actually just pushing a transpose/memory copy *into* the extern kernel.

In this case, if Inductor ended up with a bad layout.... I'd say that's Inductor's problem.

Comment (eellison) [reply torch/_inductor/decomposition.py:250]: I'm looking at a similar issue with bandwidth bound mm decomposition that ends up being extremely slow with non contiguous input. Sure, you can say it's inductor's problem, but it may still lead to regressions today
